pkg/services/ngalert/models/alert_rule.go

```diff
@@ -0,0 +1,179 @@
+	ErrAlertRuleFailedGenerateUniqueUID = errors.New("failed to generate alert rule UID")
+)
+
+type NoDataState string
+
+func (noDataState NoDataState) String() string {
+	return string(noDataState)
+}
+
+const (
+	Alerting      NoDataState = "Alerting"
+	NoData        NoDataState = "NoData"
+	KeepLastState NoDataState = "KeepLastState"
+	OK            NoDataState = "OK"
+)
+
+type ExecutionErrorState string
+
+func (executionErrorState ExecutionErrorState) String() string {
+	return string(executionErrorState)
+}
+
+const (
+	AlertingErrState      ExecutionErrorState = "Alerting"
+	KeepLastStateErrState ExecutionErrorState = "KeepLastState"
+)
+
+// AlertRule is the model for alert rules in unified alerting.
+type AlertRule struct {
+	ID              int64 `xorm:"pk autoincr 'id'"`
+	OrgID           int64 `xorm:"org_id"`
+	Title           string
+	Condition       string
+	Data            []AlertQuery
+	Updated         time.Time
+	IntervalSeconds int64
+	Version         int64
+	UID             string `xorm:"uid"`
+	NamespaceUID    string `xorm:"namespace_uid"`
+	RuleGroup       string
+	NoDataState     NoDataState
+	ExecErrState    ExecutionErrorState
+	// Receivers are used for migrating notification channels of existing alerts
+	// Receivers []string
+}
+
+// AlertRuleKey is the alert definition identifier
+type AlertRuleKey struct {
+	OrgID int64
+	UID   string
+}
+
+func (k AlertRuleKey) String() string {
+	return fmt.Sprintf("{orgID: %d, UID: %s}", k.OrgID, k.UID)
+}
+
+// GetKey returns the alert definitions identifier
+func (alertRule *AlertRule) GetKey() AlertRuleKey {
+	return AlertRuleKey{OrgID: alertRule.OrgID, UID: alertRule.UID}
+}
+
+// PreSave sets default values and loads the updated model for each alert query.
+func (alertRule *AlertRule) PreSave(timeNow func() time.Time) error {
+	for i, q := range alertRule.Data {
+		err := q.PreSave()
+		if err != nil {
+			return fmt.Errorf("invalid alert query %s: %w", q.RefID, err)
+		}
+		alertRule.Data[i] = q
+	}
+	alertRule.Updated = timeNow()
+	return nil
+}
+
+// AlertRuleVersion is the model for alert rule versions in unified alerting.
+type AlertRuleVersion struct {
+	ID               int64  `xorm:"pk autoincr 'id'"`
+	RuleOrgID        int64  `xorm:"rule_org_id"`
+	RuleUID          string `xorm:"rule_uid"`
+	RuleNamespaceUID string `xorm:"rule_namespace_uid"`
+	RuleGroup        string
+	ParentVersion    int64
+	RestoredFrom     int64
+	Version          int64
+
+	Created         time.Time
+	Title           string
+	Condition       string
+	Data            []AlertQuery
+	IntervalSeconds int64
+	NoDataState     NoDataState
+	ExecErrState    ExecutionErrorState
+	// Receivers are used for migrating notification channels of existing alerts
+	// do we still need this?
+	// Receivers []string
+}
+
+// GetAlertRuleByUIDQuery is the query for retrieving/deleting an alert rule by UID and organisation ID.
+type GetAlertRuleByUIDQuery struct {
```

Comment:
Where is this used? Is there any need for the `Result` field, i.e. can the result be returned via a normal return value instead?

Reply:
It's consumed here:
https://github.com/grafana/grafana/blob/b23bd7dbb53071c842802c691d597bfcb93aad80/pkg/services/ngalert/store/alert_rule.go#L130-L139

`SQLStore.WithDbSession` expects `dbTransactionFunc`
https://github.com/grafana/grafana/blob/5a0780801bd671515c4cc87cd6661f0c80f92843/pkg/services/sqlstore/session.go#L15
so I need the `Result` for storing and returning the retrieved alert rule to the caller.

Reply:
@papagian thanks for explaining :)